### ISSUE TITLE
th:src scripts not loaded in order

### ISSUE THREAD

**Issue by panganibanpj** [thymol/thymol.js#20]
When I do
```
<html>
  <head>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>
    <script src="https://cdn.jsdelivr.net/thymol/2.0.0/thymol.min.js"></script>
  </head>
  <body>
    <h1>start</h1>
    <script th:src="${'large.js'}"></script>
    <script th:src="${'small.js'}"></script>
  </body>
</html>
```

Where large.js is much bigger than small.js, small.js gets executed first, although it's dependent on large.js

**Comment by jjbenson** [comment]
Hi panganibanpj,

I'm sorry but I just cannot reproduce this. Maybe my test case is too simple?

I have a large.js like this:

```
window.console.log("this is large.js");
window.console.log("this is large.js");
..(196 lines omitted)..
window.console.log("this is large.js");
window.console.log("this is large.js");
function doThing() {
  window.console.log("this is small thing inside large.js");
}
```
and small.js like this:

```
window.console.log("this is small.js");
doThing();
```
The result is as expected every time.
```

large.js:1 this is large.js
large.js:2 this is large.js
large.js:3 this is large.js
....
large.js:198 this is large.js
large.js:199 this is large.js
large.js:200 this is large.js
small.js:1 this is small.js
large.js:202 this is small thing inside large.js
```
Can you please provide a specific example the illustrates the problem?

Kind Regards,

  Jim

**Comment by jjbenson** [comment]
Just FYI, if you want to use a non-default version of jquery it's best to add data-jquery-src="" to the thymol.js script tag attributes to prevent the default file from also being loaded:

e.g.
```
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>
<script src="https://cdn.jsdelivr.net/thymol/2.0.0/thymol.min.js" data-jquery-src=""></script>
```
Cheers,

  Jim

**Comment by panganibanpj** [comment]
So my large.js file was over 2Mb (~28ms download) and my small.js was about 15kb (~4ms download). My use case is that my large.js is my vendor scripts and small.js is page-specific and this particular page is very small. The files seemed to download in parallel (with large always starting to download first) but small always finished first then threw an error

**Comment by jjbenson** [comment]
I don't think that the problem is caused by the order of processing that thymol applies but by the way the browser works.

Normally you would expect both scripts to load in parallel and for "large" to run before "small" but in this case the value of "src" has been changed dynamically by thymol and that will change things.

As far as I know, when the src attribute is set, the browser will begin preparing the script for execution and as a result "small.js" find it's way at the top of the "list of scripts that will execute in order as soon as possible" before "large.js" just because it loads and parses more quickly. (See https://html.spec.whatwg.org/multipage/scripting.html#script).

I would advise you to use the JS event model and write some code to explicitly control the order of execution of large and small and make sure the cross dependencies are resolved before the code runs.

**Comment by panganibanpj** [comment]
Thanks for pointing this out! I've learned quite a bit from that and I have a better understanding of why this is happening now.
I've worked around the issue (actually didn't need the `th:src`, but tried it at first and saw this was happening). I still think it's expected behavior that the scripts get executed in the DOM order. But it doesn't look like there's a good way around the "list of scripts that will execute in order as soon as possible". Maybe just a caveat of the client-side JS
Unless you have any ideas about managing these out of the box for thymol, feel free to close this ticket.

**Comment by jjbenson** [comment]
Yes, I'm starting to think that maybe thymol should be able to do something to establish the order of execution. I **will** leave to issue open.

**Comment by panganibanpj** [comment]
Ok great! Is there something I can do to help?

**Comment by jjbenson** [comment]
As you may have guessed, this isn't a straight forward problem.
It's relatively easy to take control of the order of execution of external JavaScript on a page but more difficult to enoforce order on inline scripts, as the browser wants to execute these as soon as they are available.
Similarly, it should be possible to stop all inline scripts from executing as normal (using CSP) and to run them instead within a wrapper. Unfortunately, the CSP directive that stops JavaScript in the bodies of script tags from executing also stops all other inlined scripts such as the JavaScript content of event handler attributes.
I will continue to look into this but a fix for this issue is sure to take quite a lot of work. If I can delegate some of it, I'll be sure to let you know!
